src/Symfony/Component/HttpFoundation/Session/Storage/Handler/AbstractSessionHandler.php

```diff
@@ -0,0 +1,156 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpFoundation\Session\Storage\Handler;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+abstract class AbstractSessionHandler implements \SessionHandlerInterface, \SessionUpdateTimestampHandlerInterface
+{
+    protected $sessionName;
+
+    private $prefetchId;
+    private $prefetchData;
+    private $newSessionId;
+
+    /**
+     * @param string $sessionId
+     *
+     * @return string
+     */
+    abstract protected function doRead($sessionId);
+
+    /**
+     * @param string $sessionId
+     * @param string $data
+     *
+     * @return bool
+     */
+    abstract protected function doWrite($sessionId, $data);
+
+    /**
+     * @param string $sessionId
+     *
+     * @return bool
+     */
+    abstract protected function doDestroy($sessionId);
+
+    /**
+     * {@inheritdoc}
+     */
+    public function validateId($sessionId)
+    {
+        $this->prefetchData = $this->read($sessionId);
+        $this->prefetchId = $sessionId;
+
+        return '' !== $this->prefetchData;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function read($sessionId)
+    {
+        if (null !== $this->prefetchId) {
+            $prefetchId = $this->prefetchId;
+            $prefetchData = $this->prefetchData;
+            $this->prefetchId = $this->prefetchData = null;
+
+            if ($prefetchId === $sessionId) {
+                return $prefetchData;
+            }
+        }
+
+        $data = $this->doRead($sessionId);
+        $this->newSessionId = '' === $data ? $sessionId : null;
+
+        return $data;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function updateTimestamp($sessionId, $data)
+    {
+        return $this->write($sessionId, $data);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function write($sessionId, $data)
+    {
+        if ('' === $data) {
```

Comment:
If I am not missataken, this is where the magic happens that prevents the session from starting if you only read from it and it has not been written to before. Right?

It only tests to see if the data is an empty string. If the data is `null` or an empty array it will not be triggered. Is $data here the same as $_SESSION when you write the session, or is it a serialized verson of it?

Reply:
Just another thought: It might be a good thing to give the authors of sessonStorage classes more leeway in what counts as an empty session.

We should also have more documentation of these new things you need to consider when you write a new sessionStorage class, (but I guess that belongs in a separate documentation ticket.)

Reply:
It's always a string, pre-serialized by PHP, that's the contract and behavior of SessionHandlerInterface.
This catches the empty-session case, either no-writes, or all-deleted yes.

Reply:
Ok, thank you for answering my concern about this.